File: mobile/android/app/src/main/kotlin/com/aetherride/aetherride_mobile/LocationCorePlugin.kt
package com.aetherride.aetherride_mobile

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class LocationCorePlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var appContext: Context? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.aetherride/location_core")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "com.aetherride/location_core/fixes")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        appContext?.let { RideLocationService.stop(it) }
        RideLocationService.eventSink = null
        appContext = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val ctx = appContext
        if (ctx == null) {
            result.error("no_context", "Plugin not attached", null)
            return
        }
        when (call.method) {
            "start" -> {
                RideLocationService.start(ctx)
                result.success(true)
            }
            "stop" -> {
                RideLocationService.stop(ctx)
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        RideLocationService.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        RideLocationService.eventSink = null
    }
}

File: mobile/ios/Runner/LocationCorePlugin.swift
import CoreLocation
import Flutter
import UIKit

/// location_core iOS — CoreLocation fixes on EventChannel (parity with Android plugin).
/// Foreground tracking; no background service (iOS Background Modes optional later).
final class LocationCorePlugin: NSObject, FlutterPlugin, FlutterStreamHandler, CLLocationManagerDelegate {
  private let manager = CLLocationManager()
  private var eventSink: FlutterEventSink?
  private var methodChannel: FlutterMethodChannel?

  static func register(with registrar: FlutterPluginRegistrar) {
    let instance = LocationCorePlugin()
    let method = FlutterMethodChannel(
      name: "com.aetherride/location_core",
      binaryMessenger: registrar.messenger()
    )
    instance.methodChannel = method
    method.setMethodCallHandler(instance.handle)
    let events = FlutterEventChannel(
      name: "com.aetherride/location_core/fixes",
      binaryMessenger: registrar.messenger()
    )
    events.setStreamHandler(instance)
    instance.manager.delegate = instance
    instance.manager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
    instance.manager.distanceFilter = kCLDistanceFilterNone
    instance.manager.activityType = .fitness
    instance.manager.pausesLocationUpdatesAutomatically = false
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "start":
      let status = manager.authorizationStatus
      if status == .notDetermined {
        manager.requestWhenInUseAuthorization()
      }
      manager.startUpdatingLocation()
      result(true)
    case "stop":
      manager.stopUpdatingLocation()
      result(true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    guard let loc = locations.last, let sink = eventSink else { return }
    sink([
      "lat": loc.coordinate.latitude,
      "lng": loc.coordinate.longitude,
      "accuracyM": loc.horizontalAccuracy,
      "speedMps": max(0, loc.speed),
      "altitudeM": loc.altitude,
      "timestampMs": Int64(loc.timestamp.timeIntervalSince1970 * 1000),
    ])
  }

  func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    NSLog("location_core iOS: %@", error.localizedDescription)
  }
}
